### PR TITLE
Pre transactions

### PR DESCRIPTION
**Pull request by andreaspsson**
* Some refactoring needed to implement transactions
* version management => collision detection on commit!

### CODE REVIEW

File: datastore_entity_derives/src/lib.rs [hunk 1]
```diff
@@ -368,8 +382,8 @@ pub fn datastore_managed(input: TokenStream) -> TokenStream {
 
             fn try_from(mut entity: datastore_entity::DatastoreEntity) -> Result<Self, Self::Error> {
                 let key = entity.key();
+                let version = entity.version();
```
Comment (bes) [inline datastore_entity_derives/src/lib.rs:385]: unused?

File: src/lib.rs [hunk 2]
```diff
@@ -268,68 +235,89 @@ fn commit(
...
+    mutation.base_version = entity.version();
     let cre: CommitResponse = commit(vec![mutation], connection)?;
 
     // Assert that we have a commit result
     if let Some(results) = &cre.mutation_results {
         match results.len() {
             0 => Err(DatastoreClientError::DeleteFailed)?,
-            1 => Ok(()), // Success
+            1 => parse_mutation_result(&results[0]).map(|_| ()), // Success
             _ => Err(DatastoreClientError::AmbigiousResult)?,
```
Comment (bes) [inline src/lib.rs:318]: Ambiguous is spelled `Ambiguous`

File: tests/integration/main.rs [hunk 3]
```diff
@@ -466,5 +470,70 @@ fn test_optional_values() -> Result<(), DatastorersError> {
...
+    let prop_int_a = generate_random_int();
+    a.prop_int = prop_int_a;
+
+    let mut b = TestEntity::get_one_by_id(inserted_id, &connection)?;
+    b.prop_int = generate_random_int();
+
+    // Save the forst one => we expect success
+    a.commit(&connection)?;
+
+    // Save the second one => we expect error (collision)
```
Comment (bes) [inline tests/integration/main.rs:528]: Comment is wrong?

File: datastore_entity_derives/src/lib.rs (hunk 1 above)
Comment (andreaspsson) [reply datastore_entity_derives/src/lib.rs:385]: nope, used here `#meta_field_assignements,`

File: src/lib.rs (hunk 2 above)
Comment (andreaspsson) [reply src/lib.rs:318]: Check!

File: tests/integration/main.rs (hunk 3 above)
Comment (andreaspsson) [reply tests/integration/main.rs:528]: check!

**Pull request merged**